### DISCUSSION TITLE
Valgrind and lazy exact numbers

### DISCUSSION THREAD

**Discussion by stla** [CGAL/cgal#7025] [Q&A]
This is a follow-up of my monologues in the other posts.

I noticed that when using Valgrind, everything happens as if EPECK were EPICK.

So I did the following test. In double precision, the equality `1 - 7*0.1 == 0.3` is false.

```cpp
typedef CGAL::Lazy_exact_nt<CGAL::MP_Float> NT;

bool test() {
  NT one(1);
  NT seven(7);
  NT zeroone(0.1);
  NT x = one - seven * zeroone;
  double y = CGAL::to_double<NT>(x);
  return y == 0.3;
}
```

Results:

- without Valgrind, `test()` is `true`;
- with Valgrind, `test()` is `false`.

**Comment by lrineau** [comment]
Hi Stéphane,

I am really sorry for all the issues you have with CGAL and the test system of CRAN. I think all of those issues (including #6994), are because Valgrind does not emulate correctly the setting of the FPU rounding mode. That is documented, [in the section Limitations of its manual](https://valgrind.org/docs/manual/manual-core.html#manual-core.limits):
> For all other FP operations, only the IEEE default mode (round to nearest) is supported.

Unfortunately, the class template `CGAL::Interval_nt<P>` needs to set the rounding mode to +infinity, and that class template is at the core of the correct behavior of both `CGAL::Epick` and `CGAL::Epeck`. Under Valgrind, CGAL cannot work correctly, with the default settings.

Fortunately, there is an experimental feature of CGAL that you could use, that exists for a year in CGAL (see PR #5510) but is not well tested. If the macro `CGAL_ALWAYS_ROUND_TO_NEAREST` is defined, then `CGAL::Interval_nt<P>` is implemented differently, only using the default IEEE754 rounding mode (round to nearest). You should try to define that macro globally in your code (maybe directly in the build system), and then check if your own tests still pass. Then probably the CRAN test suite will pass more easily.

**Comment by stla** [comment]
Thanks! I will try.

**Comment by stla** [comment]
Nope :-( Unfortunately, the test above returns `false` when I define this macro.

I explained to CRAN that Valgrind does not support the CGAL exact arithmetic, I'm waiting for their reply.
